vmm/src/cpu.rs

```diff
@@ -1387,3 +1387,193 @@ impl Snapshottable for CpuManager {
+        let lint1_mode_expected = set_apic_delivery_mode(lint1, APIC_MODE_NMI);
+
+        set_lint(&vcpu).unwrap();
+
+        // Compute the value that represents LVT0 and LVT1 after set_lint.
+        let klapic_actual: LapicState = vcpu.get_lapic().unwrap();
+        let lint0_mode_actual = get_klapic_reg(&klapic_actual, APIC_LVT0);
+        let lint1_mode_actual = get_klapic_reg(&klapic_actual, APIC_LVT1);
+        assert_eq!(lint0_mode_expected, lint0_mode_actual);
+        assert_eq!(lint1_mode_expected, lint1_mode_actual);
+    }
+
+    #[test]
+    fn test_setup_fpu() {
+        #[cfg(feature = "kvm")]
+        let hv = hypervisor::kvm::KvmHypervisor::new().unwrap();
+        let hv: Arc<dyn hypervisor::Hypervisor> = Arc::new(hv);
+        let vm = hv.create_vm().expect("new VM fd creation failed");
+        let vcpu = vm.create_vcpu(0).unwrap();
+        setup_fpu(&vcpu).unwrap();
+
+        let expected_fpu: FpuState = FpuState {
+            fcw: 0x37f,
+            mxcsr: 0x1f80,
+            ..Default::default()
+        };
+        let actual_fpu: FpuState = vcpu.get_fpu().unwrap();
+        // TODO: auto-generate kvm related structures with PartialEq on.
+        assert_eq!(expected_fpu.fcw, actual_fpu.fcw);
+        // Setting the mxcsr register from FpuState inside setup_fpu does not influence anything.
+        // See 'kvm_arch_vcpu_ioctl_set_fpu' from arch/x86/kvm/x86.c.
+        // The mxcsr will stay 0 and the assert below fails. Decide whether or not we should
+        // remove it at all.
+        // assert!(expected_fpu.mxcsr == actual_fpu.mxcsr);
+    }
+
+    #[test]
+    fn test_setup_msrs() {
+        use hypervisor::arch::x86::msr_index;
+        use hypervisor::x86_64::{MsrEntries, MsrEntry};
+
+        #[cfg(feature = "kvm")]
+        let hv = hypervisor::kvm::KvmHypervisor::new().unwrap();
+        let hv: Arc<dyn hypervisor::Hypervisor> = Arc::new(hv);
+        let vm = hv.create_vm().expect("new VM fd creation failed");
+        let vcpu = vm.create_vcpu(0).unwrap();
+        setup_msrs(&vcpu).unwrap();
+
+        // This test will check against the last MSR entry configured (the tenth one).
+        // See create_msr_entries for details.
+        let mut msrs = MsrEntries::from_entries(&[MsrEntry {
+            index: msr_index::MSR_IA32_MISC_ENABLE,
+            ..Default::default()
+        }]);
+
+        // get_msrs returns the number of msrs that it succeed in reading. We only want to read 1
+        // in this test case scenario.
+        let read_msrs = vcpu.get_msrs(&mut msrs).unwrap();
+        assert_eq!(read_msrs, 1);
+
+        // Official entries that were setup when we did setup_msrs. We need to assert that the
+        // tenth one (i.e the one with index msr_index::MSR_IA32_MISC_ENABLE has the data we
+        // expect.
+        let entry_vec = hypervisor::x86_64::boot_msr_entries();
+        assert_eq!(entry_vec.as_slice()[9], msrs.as_slice()[0]);
+    }
```

Comment:
These tests are a lot more hypervisor than vmm related. `test_setup_msrs` for example exclusively tests the hypervisor API/traits.
So I think most of them should move as hypervisor unit tests instead. And if that means including a few architecture specific definitions (no `arch` dependencies, don't get me wrong) to the hypervisor crate, this is fine with me.

Reply:
I was about to mention that some functions' implementations live in `arch` crate, but I think those functions are just wrappers around trait methods in the `hypervisor` crate.

I will have a closer look.

Reply:
Alright. There are some non-trivial functions from `arch` crate that are used in test cases, say, `setup_page_tables` and `configure_segments_and_sregs`. What do you want to do with those functions?

Reply:
Ack.